.github/workflows/release.yml

```diff
@@ -46,96 +88,117 @@ jobs:
+    needs: [check, build]
+    if: needs.check.outputs.projects != '[]'
+    name: Release [publish=${{ github.event_name != 'pull_request' }}]
     permissions:
       # This enables the job to create and/or update GitHub releases
       contents: write
     runs-on: ubuntu-latest
-    needs: [build]
+    strategy:
+      fail-fast: false
+      matrix:
+        project: ${{ fromJSON(needs.check.outputs.projects) }}
     env:
-      PUBLISH: ${{ startsWith(inputs.ref || github.ref, 'refs/tags/') }}
+      PUBLISH: ${{ github.event_name != 'pull_request' }}
     steps:
-      - uses: actions/checkout@v4
-        with:
-          path: actions
       - uses: actions/checkout@v4
         with:
           submodules: 'recursive'
           fetch-depth: 0
-          path: lotus
-          ref: ${{ inputs.ref || github.ref }}
       - uses: actions/download-artifact@v4
         with:
-          name: lotus-Linux-X64
+          name: lotus-${{ matrix.project }}-Linux-X64
           path: linux_amd64_v1
       - uses: actions/download-artifact@v4
         with:
-          name: lotus-macOS-X64
+          name: lotus-${{ matrix.project }}-macOS-X64
           path: darwin_amd64_v1
       - uses: actions/download-artifact@v4
         with:
-          name: lotus-macOS-ARM64
+          name: lotus-${{ matrix.project }}-macOS-ARM64
           path: darwin_arm64
-      - uses: ./actions/.github/actions/install-go
-        with:
-          working-directory: lotus
+      - uses: ./.github/actions/install-go
       - uses: ipfs/download-ipfs-distribution-action@v1
         with:
           name: kubo
           version: v0.16.0
+      - id: meta
+        run: |
+          version="$(go run cmd/release/main.go --json ${{ matrix.project }} version | jq -r .msg)"
+          if [[ "${{ matrix.project}}" == "miner" ]]; then
+            tag="miner/$version"
+            latest="$(git tag -l 'miner/v*' | sort -V | tail -n1)"
+          elif [[ "${{ matrix.project}}" == "node" ]]; then
+            tag="$version"
+            latest="$(git tag -l 'v*' | sort -V | tail -n1)"
+          else
+            echo "::error title=Unknown Project::${{ matrix.project }}"
+            exit 1
+          fi
+          latest="$(echo -e "$latest\n$version" | sort -V | tail -n1)"
+          echo "version=$version" | tee -a $GITHUB_OUTPUT
+          echo "latest=$latest" | tee -a $GITHUB_OUTPUT
+          echo "tag=$tag" | tee -a $GITHUB_OUTPUT
+      - env:
+          expected: ${{ steps.meta.outputs.version }}
+        run: |
+          for dir in linux_amd64_v1 darwin_amd64_v1 darwin_arm64; do
+            for bin in lotus lotus-miner lotus-worker; do
+              if [[ -f $dir/$bin ]]; then
+                actual=$(./$dir/$bin --version)
+                if [[ "$actual" != "$expected" ]]; then
+                  echo "::error title=Version Mismatch::Expected $expected, got $actual ($dir/$bin)"
```

Comment:
nice, this is new, right? have we tested this?

Reply:
Yes, now we did - I forgot the binaries are for different archs. I now moved the tests back to the build jobs.